HeinOnline.js

```diff
@@ -0,0 +1,224 @@
+	var query = url.replace(/.*?\?/, "");
+	query = query.split("&");
+	for (var i=0,ilen=query.length;i<ilen;i++) {
+		var pair = query[i].split("=");
+		ret[pair[0]] = pair[1];
+	}
+	return ret;
+}
+
+// Remangle HeinOnline RIS to form expected by translator
+function fixRIS (ris) {
+	var lines = ris.split("\n");
+	for (var i=0,ilen=lines.length;i<ilen;i++) {
+		lines[i] = lines[i].replace(/^([A-Z0-9][A-Z0-9]) /, "$1  - ").replace(/^VO  /, "VL  ").replace(/^YR  /, "DA  ").replace(/^OP  /, "EP  ");
+	}
+	ris = lines.join("\n");
+	return ris;
+}
+
+// Not all pages have a downloadable PDF
+function translateRIS(ris, pdfURL) {
+	var trans = Zotero.loadTranslator('import');
+	trans.setTranslator('32d59d2d-b65a-4da4-b0a3-bdd3cfb979e7');//https://github.com/zotero/translators/blob/master/RIS.js
+	trans.setString(ris);
+	trans.setHandler('itemDone', function (obj, item) {
+		if (pdfURL) {
+			item.attachments = [{
+				title: "Full Text PDF",
+				url: pdfURL,
+				mimeType: "application/pdf"
+			}];
+			
+		}
+		item.complete();
+	});
+	trans.translate();
+}
+
+function translateCOinS(COinS) {
+	var item = new Zotero.Item();
+	Zotero.Utilities.parseContextObject(COinS, item);
+	item.complete();
+}
+
+// Build URL for RIS, and for PDF if available
+function scrapePage(doc, url) {
+	var risPopupURL = getXPathStr("href", doc, '//form[@id="pagepicker"]//a[contains(@href, "PrintRequest")][1]');
+	if (risPopupURL) {
+		// If page has RIS, use that
+		var docParams = extractQueryValues(risPopupURL);
+		var risURL = docParams.base 
+			+ "CitationFile?kind=ris&handle=" + docParams.handle 
+			+ "&div=" + docParams.div 
+			+ "&id=" + docParams.id 
+			+ "&base=js";
+		var pdfPageURL = doc.getElementsByClassName("updatediv");
+		ZU.doGet(risURL, function(ris) {
+			ris = fixRIS(ris);
+			if (pdfPageURL) {
+				ZU.doGet(pdfPageURL, function(pdfPage) {
+					// Call to pdfPageURL prepares PDF for download via META refresh URL
+					var pdfURL = null;
+					var m = pdfPage.match(/<META.*URL=([^"]+)/);
+					if (m) {
+						var pdfURL = m[1];
+					}
+					translateRIS(ris, pdfURL);
+				} , null );
+			} else {
+				translateRIS(ris);
+			}
+		} , null );
+	} else {
+		// No RIS available in page, try COinS
+		var COinS = getXPathStr("title", doc, '//span[contains(@class, "Z3988")]');
+		if (COinS) {
+			translateCOinS(COinS);
+		}
+	}
+}
+
+/*
+	*********
+	** API **
+	*********
+*/
+
+function detectWeb (doc, url) {
+	if (url.indexOf("/LuceneSearch?") > -1) {
+		if (getSearchResults(doc)) {
+			return "multiple";
+		}
+	} else {
+		return "journalArticle";
+	}
+	return false;
+}
+
+function doWeb (doc) {
+	if (detectWeb(doc) === "multiple") {
+		Zotero.selectItems(getSearchResults(doc), function (items) {
+
+			if (!items) {
+				return true;
+			}
+			var urls = [];
+			for (var i in items) {
+				urls.push(i);
+			}
+
+			ZU.processDocuments(urls, scrapePage);
+		});
+	} else {
+		scrapePage(doc, url);
+	}
+}
+
+/** BEGIN TEST CASES **/
+var testCases = [
```

Comment:
as per aurimas, more tests cases would be great (if you're not using scaffold alredy, you'll definitely want to for that purpose, since it creates test JSON automatically.)]
https://www.zotero.org/support/dev/translators/scaffold